device/core/src/sak_authentication_provider.ts

```diff
@@ -81,11 +79,7 @@ export class SharedAccessKeyAuthenticationProvider extends EventEmitter implemen
     return (this._currentTokenExpiryTimeInSeconds - currentTimeInSeconds) < this._tokenRenewalMarginInSeconds;
   }
 
-  private _renewToken(callback?: (err: Error, credentials?: TransportConfig) => void): void {
-    if (this._renewalTimeout) {
-      clearTimeout(this._renewalTimeout);
-    }
-
```

Comment:
race condition?  If you call getDeviceCredentials right before the timer fires, could the timer fire during the signature operation?

Reply:
You could but in theory these if() statements to clear/restart the timer should protect us from this.

---
In reply to: [197949683](https://github.com/Azure/azure-iot-sdk-node/pull/330#discussion_r197949683) [](ancestors = 197949683)

Reply:
It's a small window, but it's there .
user calls getDeviceCredentials, getDeviceCredentials calls _renewToken,  after _renewToken returns, the timer is cleared and restarted.
However, if the timer goes off during _sign (the most likely place in a small window), _expiryTimerHandler will call _renewToken and schedule a new timer.
There's nothing stopping the expiry handler from renewing the token even though the user flow is already renewing it.
(this is all very unlikely and only matters if any code in this flow yields and lets the timer fire)





---
In reply to: [197950960](https://github.com/Azure/azure-iot-sdk-node/pull/330#discussion_r197950960) [](ancestors = 197950960,197949683)

Reply:
oh I see - this being said, whichever piece of code finishes last will restart the timer, right? so it's a race but it doesn't block us. in any case, it feels like maybe we need a bit more state management now that this thing is async.

---
In reply to: [197997041](https://github.com/Azure/azure-iot-sdk-node/pull/330#discussion_r197997041) [](ancestors = 197997041,197950960,197949683)

Reply:
I think you end up with a double-renew and then the second piece of code schedules the timer.   Not a big deal.  

---
In reply to: [197997354](https://github.com/Azure/azure-iot-sdk-node/pull/330#discussion_r197997354) [](ancestors = 197997354,197997041,197950960,197949683)

Reply:
I could cancel the timer as soon as I call getDeviceCredentials, that way I'm sure a signing operation isn't triggered by a timer, but I don't know if I can do that the other way around. 

---
In reply to: [197997631](https://github.com/Azure/azure-iot-sdk-node/pull/330#discussion_r197997631) [](ancestors = 197997631,197997354,197997041,197950960,197949683)